src/coreclr/src/vm/eventtrace.cpp

```diff
@@ -3498,75 +3498,87 @@ BOOL ETW::TypeSystemLog::AddTypeToGlobalCacheIfNotExists(TypeHandle th, BOOL * p
                 return fSucceeded;
             }
         }
+    }
+
+    // Step 1: go from LoaderModule to hash of types.
+    Module * pLoaderModule = th.GetLoaderModule();
+    _ASSERTE(pLoaderModule != NULL);
+    LoggedTypesFromModule * pLoggedTypesFromModule = nullptr;
 
-        // Step 1: go from LoaderModule to hash of types.
+    {
+        CrstHolder _crst(GetHashCrst());
+        
+        pLoggedTypesFromModule = s_pAllLoggedTypes->allLoggedTypesHash.Lookup(pLoaderModule);
+    }
 
-        Module * pLoaderModule = th.GetLoaderModule();
-        _ASSERTE(pLoaderModule != NULL);
-        LoggedTypesFromModule * pLoggedTypesFromModule = s_pAllLoggedTypes->allLoggedTypesHash.Lookup(pLoaderModule);
+    if (pLoggedTypesFromModule == NULL)
+    {
+        pLoggedTypesFromModule = new (nothrow) LoggedTypesFromModule(pLoaderModule);
         if (pLoggedTypesFromModule == NULL)
         {
-            pLoggedTypesFromModule = new (nothrow) LoggedTypesFromModule(pLoaderModule);
-            if (pLoggedTypesFromModule == NULL)
-            {
-                // out of memory.  Bail on ETW stuff
-                *pfCreatedNew = FALSE;
-                return fSucceeded;
-            }
-
-            fSucceeded = FALSE;
-            EX_TRY
-            {
-                s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule);
-                fSucceeded = TRUE;
-            }
-            EX_CATCH
-            {
-                fSucceeded = FALSE;
-            }
-            EX_END_CATCH(RethrowCorruptingExceptions);
-            if (!fSucceeded)
-            {
-                *pfCreatedNew = FALSE;
-                return fSucceeded;
-            }
+            // out of memory.  Bail on ETW stuff
+            *pfCreatedNew = FALSE;
+            return fSucceeded;
         }
+    }
 
-        // Step 2: From hash of types, see if our TypeHandle is there already
-        TypeLoggingInfo typeLoggingInfoPreexisting = pLoggedTypesFromModule->loggedTypesFromModuleHash.Lookup(th);
+    fSucceeded = FALSE;
+    EX_TRY
+    {
+        CrstHolder _crst(GetHashCrst());
+        s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule);
```

Comment:
This isn't correct. If in the intervening time between a allLoggedTypesHash.Lookup call above, and this Add, another thread may have added to the allLoggedTypesHash. In that case, this Add will result in multiple entries in the SHash. That itself isn't a problem as SHash supports that, and for the ETW event stream, multiple BulkType events emitted won't ever cause a failure, but the logic in TypeSystemLog::OnModuleUnload won't then correctly delete all of the LoggedTypesFromModule objects, and if a new module is loaded as the same memory address as the module that wasn't properly cleaned up, then incorrect behavior may be observed.

Also, it appears that this unconditionally adds the pLoggedTypesFromModule to the hash even if it was found via Lookup above. That should not be done with a SHash based hash table.

Instead the logic should use a NewHolder to hold the newly allocated LoggedTypesFromModule, and then take the critical section, and attempt to add the newly allocated LoggedTypesFromModule object into the hashtable, by taking a lock, attempting to Lookup, if the lookup fails, then Adding, and releasing the NewHolder.